### PR TITLE
feat(core,developer): support normalization=disabled 🙀

### PR DESCRIPTION
**Pull request by srl295**
Fixes: #10554

- updates on the developer and core side
- 'meta' section (both sides) has a getter to return status
- .. plus tests of course
- also, dropped unneeded assert() calls and generally restructured ldml_processor::process_output

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/types/src/kmx/kmx-plus.ts [hunk 1]
```diff
@@ -75,6 +75,11 @@ export class Meta extends Section {
   indicator: StrsItem;
   version: StrsItem; // semver version string, defaults to "0"
   settings: KeyboardSettings;
+
+  /** convenience for checking settings */
+  get normalizionDisabled() {
```
Comment (mcdurdin) [inline common/web/types/src/kmx/kmx-plus.ts:80]:
```suggestion
  get normalizationDisabled() {
```

File: common/web/types/src/kmx/kmx-plus.ts [hunk 2]
```diff
@@ -205,11 +210,15 @@ export class Strs extends Section {
...
     // nfd
     if (opts?.nfd) {
-      if (opts?.markers) {
-        s = MarkerParser.nfd_markers(s, false);
-      } else {
-        s = s.normalize("NFD");
-      }
+      if (!sections.meta) {
+        throw Error(`Internal Error: need 'meta' section to check normalization mode.`);
+      } else if (!sections.meta.normalizionDisabled) {
```
Comment (mcdurdin) [inline common/web/types/src/kmx/kmx-plus.ts:215]:
```suggestion
      } else if (!sections.meta.normalatizionDisabled) {
```

File: common/web/types/src/kmx/kmx-plus.ts [hunk 3]
```diff
@@ -205,11 +210,15 @@ export class Strs extends Section {
     }
     // nfd
     if (opts?.nfd) {
-      if (opts?.markers) {
-        s = MarkerParser.nfd_markers(s, false);
-      } else {
-        s = s.normalize("NFD");
-      }
+      if (!sections.meta) {
+        throw Error(`Internal Error: need 'meta' section to check normalization mode.`);
```
Comment (mcdurdin) [inline common/web/types/src/kmx/kmx-plus.ts:214]: If no `<meta>` section is included, does it default to normalizationDisabled = false?

File: core/src/ldml/ldml_processor.cpp [hunk 4]
```diff
@@ -298,19 +298,26 @@ size_t ldml_processor::process_output(ldml_event_state &ldml_state, const std::u
   //  The normalize functions have assert and Debuglog at the bottom.
   //  so we do not need to assert the status here unless we're going to do something
   //  different with control flow.
-  (void)ldml::normalize_nfd_markers(nfd_str);
-
+  if (!normalization_disabled) {
+    auto norm_ok = ldml::normalize_nfd_markers(nfd_str);
+    assert(norm_ok);
```
Comment (mcdurdin) [inline core/src/ldml/ldml_processor.cpp:303]:
```suggestion
    assert(ldml::normalize_nfd_markers(nfd_str));
```

Can we do this and reduce the noise?

File: core/src/ldml/ldml_processor.cpp [hunk 5]
```diff
@@ -298,19 +298,26 @@ size_t ldml_processor::process_output(ldml_event_state &ldml_state, const std::u
   //  The normalize functions have assert and Debuglog at the bottom.
   //  so we do not need to assert the status here unless we're going to do something
   //  different with control flow.
-  (void)ldml::normalize_nfd_markers(nfd_str);
-
+  if (!normalization_disabled) {
```
Comment (mcdurdin) [inline core/src/ldml/ldml_processor.cpp:301]: This is a repeated pattern, should we be extracting it out?

File: core/src/ldml/ldml_processor.cpp [hunk 6]
```diff
@@ -330,14 +337,18 @@ size_t ldml_processor::process_output(ldml_event_state &ldml_state, const std::u
 
   // drop last 'matchedContext':
   ctxtstr.resize(ctxtstr.length() - matchedContext);
-  ctxtstr.append(outputString); // TODO-LDML: should be able to do a normalization-safe append here.
-  (void)ldml::normalize_nfd_markers(ctxtstr);
-
+  ctxtstr.append(outputString); // TODO-LDML: should be able to do a normalization-safe append here.'
+  if (!normalization_disabled) {
+    (void)ldml::normalize_nfd_markers(ctxtstr);
```
Comment (mcdurdin) [inline core/src/ldml/ldml_processor.cpp:342]: Why no assertion on this instance? Seems inconsistent (See extracting-it-out comment above)

File: core/src/ldml/ldml_transforms.cpp [hunk 7]
```diff
@@ -512,9 +514,11 @@ transform_entry::init() {
   }
   // TODO-LDML: if we have mapFrom, may need to do other processing.
   std::u32string from2 = fFrom;
-  normalize_nfd_markers(from2, regex_sentinel);
+  if (!normalization_disabled) {
+    // normalize, including markers, for regex
+    normalize_nfd_markers(from2, regex_sentinel);
```
Comment (mcdurdin) [inline core/src/ldml/ldml_transforms.cpp:519]: no assertion?

File: core/src/ldml/ldml_transforms.cpp [hunk 8]
```diff
@@ -637,7 +641,7 @@ transform_entry::apply(const std::u32string &input, std::u32string &output) cons
     }
     output.assign(s.get(), out32len);
     // NOW do a marker-safe normalize
-    if (!normalize_nfd_markers(output)) {
+    if (!normalization_disabled && !normalize_nfd_markers(output)) {
       DebugLog("normalize_nfd_markers(output) failed");
       return 0; // TODO-LDML: normalization failed.
     }
```
Comment (mcdurdin) [inline core/src/ldml/ldml_transforms.cpp:647]: No assertion?

**Review by mcdurdin — APPROVED** [review]
LGTM

There are still many TODO-LDML comments, some of which are a concern, relating to normalization. Can we review and make sure there are no gaps before beta?

File: core/src/ldml/ldml_processor.cpp (hunk 4 above)
Comment (srl295) [reply core/src/ldml/ldml_processor.cpp:303]: @mcdurdin except that the `assert()` is not called in production, right? maybe we need a new construct.  

`#define assert(x)` is the usual idiom for production mode.

File: core/src/ldml/ldml_processor.cpp (hunk 6 above)
Comment (srl295) [reply core/src/ldml/ldml_processor.cpp:342]: @mcdurdin perhaps an inline member would make sense.

File: core/src/ldml/ldml_processor.cpp (hunk 4 above)
Comment (mcdurdin) [reply core/src/ldml/ldml_processor.cpp:303]: Ah of course. My bad ... that could be ugly.

`assert(x)` is dangerous like that... That does mean that `norm_ok` is declared and unused in production code which is possibly a separate painful issue.

Comment (srl295) [reply core/src/ldml/ldml_processor.cpp:303]: if the code was extracted out, then the pain would be only in one place.

But actually.  the inner level `normalize()` function *already* asserts.  So one thing I could do is just take out all of these other asserts.

Comment (srl295) [reply core/src/ldml/ldml_processor.cpp:303]: hah, and i had added this comment three weeks ago:

```
size_t ldml_processor::process_output(ldml_event_state &ldml_state, const std::u32string &str, ldml::transforms *with_transforms) const {
  std::u32string nfd_str = str;
  // Note:
  //  The normalize functions have assert and Debuglog at the bottom.
  //  so we do not need to assert the status here unless we're going to do something
  //  different with control flow.
```

**Review by mcdurdin — APPROVED** [review]

**Comment by srl295** [comment]
> LGTM
> 
> 
> 
> There are still many TODO-LDML comments, some of which are a concern, relating to normalization. Can we review and make sure there are no gaps before beta?

I think i took out lots of todos- in any event yes will remove as discussed

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.262-alpha](https://keyman.com/downloads/releases/alpha/17.0.262)

**Pull request merged**